utils/clients/sftp_client.py

```diff
@@ -0,0 +1,102 @@
+# ############################################################################### #
+# Autoreduction Repository : https://github.com/ISISScientificComputing/autoreduce
+#
+# Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI
+# SPDX - License - Identifier: GPL-3.0-or-later
+# ############################################################################### #
+"""
+Client class for retrieving files via SFTP from servers (e.g. CEPH)
+"""
+import os.path
+import logging
+import pysftp
+from utils.clients.abstract_client import AbstractClient
+from utils.clients.connection_exception import ConnectionException
+from utils.test_settings import SFTP_SETTINGS
+from utils.project.structure import get_log_file
+from utils.project.static_content import LOG_FORMAT
+
+
+logging.basicConfig(filename=get_log_file('sftp_client.log'), level=logging.INFO,
+                    format=LOG_FORMAT)
+
+
+class SFTPClient(AbstractClient):
+    """
+    This class allows files to be retrieved via SFTP from servers (e.g. CEPH)
+    """
+    def __init__(self, credentials=None):
+        if not credentials:
+            credentials = SFTP_SETTINGS
+        super(SFTPClient, self).__init__(credentials)
+        self._connection = None
+
+    def connect(self):
+        """
+        Create the connection to the SFTP server
+        :return: connection object
+        """
+        if self._connection is None:
+            cnopts = pysftp.CnOpts()
+            cnopts.hostkeys = None
+            self._connection = pysftp.Connection(host=self.credentials.host,
+                                                 username=self.credentials.username,
+                                                 password=self.credentials.password,
+                                                 port=int(self.credentials.port),
+                                                 cnopts=cnopts)
+        self._test_connection()
+
+        return self._connection
+
+    def disconnect(self):
+        """
+        Disconnect from the SFTP server
+        """
+        if self._connection is not None:
+            self._connection.close()
+        self._connection = None
+
+    def _test_connection(self):
+        """
+        Test whether there is a connection to the SFTP server
+        :return: True if there is a connection
+        """
+
+        try:
+            self._connection.pwd
+        except AttributeError:
+            raise ConnectionException("SFTP")
+        return True
+
+    def retrieve(self, server_path, local_path=None, override=True):
+        """
+        Retrieves file from the given server_path and downloads it to the given local_path
```

Comment:
again `:raises:` tag. This is not used in the rest of the code base currently but is something we are looking to add as we go